### PR TITLE
Fixes #33044 - re-use existing publication on sync

### PR DESCRIPTION
**Pull request by jlsherrill**
As part of pulp 3.14 publications can now be created during sync time
in this case, if a publication appears, we should re-use it for
distribution instead of regenerating it.

### CODE REVIEW

**Comment by jlsherrill** [comment]
Things to test:

1) Yum syncs with mirror = true
2) Yum syncs without mirror = true3
3) file repo syncs
4) capsule Syncs with pulp 3.14
5) capsule syncs with pulp 3.11 or earlier

**Comment by jlsherrill** [comment]
there are new tests to write and old tests to fix, will do that next

**Comment by ianballou** [comment]
Weird, having trouble syncing kickstart repos. Probably unrelated to your PR, but I was hoping to test them since there was that publication bug.

```
{"pulp_tasks"=>
  [{"pulp_href"=>"/pulp/api/v3/tasks/04975f0a-4ce4-4a16-9c84-28473faedc63/",
    "pulp_created"=>"2021-07-16T21:11:57.441+00:00",
    "state"=>"failed",
    "name"=>"pulp_rpm.app.tasks.synchronizing.synchronize",
    "logging_cid"=>"37868873-0f53-47c8-b70e-c031a6f83b18",
    "started_at"=>"2021-07-16T21:11:57.493+00:00",
    "finished_at"=>"2021-07-16T21:12:03.416+00:00",
    "error"=>
     {"traceback"=>
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/tasking/pulpcore_worker.py\", line 266, in _perform_task\n" +
       "    result = func(*args, **kwargs)\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulp_rpm/app/tasks/synchronizing.py\", line 422, in synchronize\n" +
       "    version = dv.create()\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/plugin/stages/declarative_version.py\", line 151, in create\n" +
       "    loop.run_until_complete(pipeline)\n" +
       "  File \"/usr/lib64/python3.6/asyncio/base_events.py\", line 484, in run_until_complete\n" +
       "    return future.result()\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/plugin/stages/api.py\", line 225, in create_pipeline\n" +
       "    await asyncio.gather(*futures)\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/plugin/stages/api.py\", line 43, in __call__\n" +
       "    await self.run()\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulp_rpm/app/tasks/synchronizing.py\", line 693, in run\n" +
       "    result = await downloader.run()\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/download/http.py\", line 258, in run\n" +
       "    return await download_wrapper()\n" +
       "  File \"/usr/lib/python3.6/site-packages/backoff/_async.py\", line 133, in retry\n" +
       "    ret = await target(*args, **kwargs)\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/download/http.py\", line 256, in download_wrapper\n" +
       "    return await self._run(extra_data=extra_data)\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulp_rpm/app/downloaders.py\", line 93, in _run\n" +
       "    to_return = await self._handle_response(response)\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/download/http.py\", line 212, in _handle_response\n" +
       "    await self.finalize()\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/download/base.py\", line 162, in finalize\n" +
       "    self.validate_digests()\n" +
       "  File \"/usr/lib/python3.6/site-packages/pulpcore/download/base.py\", line 211, in validate_digests\n" +
       "    if expected_digest != self._digests[algorithm].hexdigest():\n",
      "description"=>"'md5'"},
    "worker"=>"/pulp/api/v3/workers/1d3c18e0-3be7-4443-8288-27349692ac5a/",
    "child_tasks"=>[],
    "progress_reports"=>
     [{"message"=>"Downloading Metadata Files",
       "code"=>"sync.downloading.metadata",
       "state"=>"failed",
       "done"=>8},
      {"message"=>"Downloading Artifacts",
       "code"=>"sync.downloading.artifacts",
       "state"=>"canceled",
       "done"=>0},
      {"message"=>"Associating Content",
       "code"=>"associating.content",
       "state"=>"canceled",
       "done"=>0}],
    "created_resources"=>[],
    "reserved_resources_record"=>
     ["/pulp/api/v3/remotes/rpm/rpm/5b93d748-0e40-40a7-aae8-eb1a3e808b00/",
      "/pulp/api/v3/repositories/rpm/rpm/db27d244-172f-4145-8ebc-3a620a971bc0/"]}],
 "create_version"=>true,
 "task_groups"=>[],
 "poll_attempts"=>{"total"=>8, "failed"=>1}}
```

**Comment by ianballou** [comment]
Also probably unrelated to this PR, but Pulp 2 proxy syncing is having issues:

```
---
- "/home/vagrant/katello/app/services/cert/certs.rb:15:in `ssl_client_cert'"
- "/home/vagrant/katello/app/services/katello/pulp/server.rb:20:in `config'"
- "/home/vagrant/katello/app/models/katello/concerns/smart_proxy_extensions.rb:127:in
  `pulp_api'"
- "/home/vagrant/katello/app/lib/actions/pulp/abstract.rb:30:in `capsule_content'"
- "/home/vagrant/katello/app/lib/actions/pulp/abstract.rb:14:in `pulp_resources'"
- "/home/vagrant/katello/app/lib/actions/pulp/consumer/unassociate_units.rb:15:in
  `invoke_external_task'"
```

Pulp 3 proxy syncing is working fine so far.

**Comment by jlsherrill** [comment]
Filed this bug for the rhel 8.4 ks repo sync error:  https://pulp.plan.io/issues/9111   enabling md5 checksums in pulp3 works around it

Pulp 2 capsule syncs aren't really supported in master anymore

**Comment by ianballou** [comment]
Okay, I got through the  whole testing checklist. Hit this bug (https://pulp.plan.io/issues/8807) on Pulpcore 3.11 proxy syncing, but removing the kickstart repo made it work.

**Comment by ianballou** [comment]
Looks like there are more VCR issues.

File: app/services/katello/pulp3/task.rb [hunk 1]
```diff
@@ -51,6 +51,22 @@ def initialize(smart_proxy, data)
...
+        version_hrefs.last
+      end
+
+      def self.publication_href(tasks)
+        tasks = [tasks] unless tasks.is_a?(Array)
+        publication_hrefs = tasks.map { |task| task[:created_resources] }.flatten
+        publication_hrefs = publication_hrefs.select { |href| ::Katello::Pulp3::Repository.publication_href?(href) }
+        Rails.logger.error("Got multiple publication hrefs for pulp task: #{tasks}") if publication_hrefs.length > 2
+        publication_hrefs.last #return the last href to workaround https://pulp.plan.io/issues/9098
+      end
```
Comment (jlsherrill) [inline app/services/katello/pulp3/task.rb:68]: changed these method to take in multiple tasks and consider all of them, apparently as part of unit copy it was passing in multiple tasks and the first one had no version info

**Review by ianballou — APPROVED** [review]
Looking fine to me! Just need the tests to pass.

**Pull request merged**